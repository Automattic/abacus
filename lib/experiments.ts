import {
  AnalysisStrategy,
  ExperimentFull,
  ExperimentFullNew,
  ExperimentFullNormalizedData,
  Platform,
  Variation,
} from './schemas'

/**
 * Return the deployed variation if one has been selected, otherwise `null`.
 *
 * @throws {Error} If a `deployedVariationId` is set but cannot be found in the
 *   variations.
 */
export function getDeployedVariation(normalizedExperimentData: ExperimentFullNormalizedData): null | Variation {
  const normalizedExperiment = normalizedExperimentData.entities.experiments[normalizedExperimentData.result]

  if (typeof normalizedExperiment.deployedVariationId !== 'number') {
    return null
  }

  const deployedVariation = normalizedExperimentData.entities.variations[normalizedExperiment.deployedVariationId]

  if (!deployedVariation) {
    throw new Error(
      `Failed to resolve the deployed variation with ID ${normalizedExperiment.deployedVariationId} for experiment with ID ${normalizedExperiment.experimentId}.`,
    )
  }

  return deployedVariation
}

/**
 * Return the primary metric assignment ID for this experiment if one exists.
 */
export function getPrimaryMetricAssignmentId(experiment: ExperimentFull): number | null {
  return experiment.metricAssignments.find((metricAssignment) => metricAssignment.isPrimary)?.metricAssignmentId ?? null
}

/**
 * Determines whether conclusion data has been entered for this experiment.
 */
export function hasConclusionData(experiment: ExperimentFull): boolean {
  return !!experiment.endReason || !!experiment.conclusionUrl || typeof experiment.deployedVariationId === 'number'
}

/**
 * Return this experiment's default analysis strategy, which depends on the existence of exposureEvents.
 */
export function getDefaultAnalysisStrategy(experiment: ExperimentFull) {
  return experiment.exposureEvents ? AnalysisStrategy.PpNaive : AnalysisStrategy.MittNoSpammersNoCrossovers
}

export function createNewExperiment(): Partial<ExperimentFullNew> {
  return {
    platform: Platform.Wpcom,
    metricAssignments: [],
    segmentAssignments: [],
    variations: [
      { name: 'control', isDefault: true, allocatedPercentage: 50 },
      { name: 'treatment', isDefault: false, allocatedPercentage: 50 },
    ],
  }
}

export const PlatformToHuman: Record<Platform, string> = {
  [Platform.Wpcom]: 'WordPress.com',
  [Platform.Calypso]: 'Calypso',
}
