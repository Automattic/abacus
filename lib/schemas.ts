// Schema documentation lives at:
// https://app.swaggerhub.com/apis/yanir/experiments/0.1.0

import * as yup from 'yup'

const idSchema = yup.number().integer().positive()

export const metricAssignmentSchema = yup
  .object({
    metricAssignmentId: idSchema.defined(),
    attributionWindowSeconds: yup.number().integer().positive().defined(),
    changeExpected: yup.bool().defined(),
    experimentId: idSchema.defined(),
    isPrimary: yup.bool().defined(),
    metricId: idSchema.defined(),
    minDifference: yup.number().defined(),
  })
  .defined()
  .camelCase()

export type MetricAssignment = yup.InferType<typeof metricAssignmentSchema>

export const segmentAssignmentSchema = yup
  .object({
    segmentAssignmentId: idSchema.defined(),
    experimentId: idSchema.defined(),
    segmentId: idSchema.defined(),
    isExcluded: yup.bool().defined(),
  })
  .defined()
  .camelCase()

export type SegmentAssignment = yup.InferType<typeof segmentAssignmentSchema>

export const variationSchema = yup
  .object({
    variationId: idSchema.defined(),
    experimentId: idSchema.defined(),
    name: yup.string().max(128).defined(),
    isDefault: yup.bool().defined(),
    allocatedPercentage: yup.number().integer().min(1).max(99).defined(),
  })
  .defined()
  .camelCase()

export type Variation = yup.InferType<typeof variationSchema>

/**
 * The platform where the experiment is running.
 * - `calypso`: The experiment is being run on the front-end Calypso interface,
 *   WordPress.com. Account sign-up and site management design experiments are
 *   likely run here.
 * - `wpcom`: The experiment is being run on the back-end, like APIs which are
 *   usually written in PHP. Email and landing pages experiments are likely run here.
 */
export enum Platform {
  Calypso = 'calypso',
  Wpcom = 'wpcom',
}

/**
 * The status of an experiment.
 */
export enum Status {
  Staging = 'staging',
  Running = 'running',
  Completed = 'completed',
  Disabled = 'disabled',
}

export const experimentBareSchema = yup
  .object({
    experimentId: idSchema.defined(),
    name: yup.string().max(128).defined(),
    startDatetime: yup.date().defined(),
    endDatetime: yup.date().defined(),
    status: yup.string().oneOf(Object.values(Status)).defined(),
    platform: yup.string().oneOf(Object.values(Platform)).defined(),
    ownerLogin: yup.string().defined(),
  })
  .defined()
  .camelCase()

export type ExperimentBare = yup.InferType<typeof experimentBareSchema>

export const experimentFullSchema = experimentBareSchema
  .shape({
    description: yup.string().defined(),
    existingUsersAllowed: yup.boolean().defined(),
    p2Url: yup.string().url().defined(),
    endReason: yup.string().defined().nullable(),
    conclusionUrl: yup.string().url().defined().nullable(),
    deployedVariationId: idSchema.defined().nullable(),
    metricAssignments: yup.array(metricAssignmentSchema).defined(),
    segmentAssignments: yup.array(segmentAssignmentSchema).defined(),
    variations: yup.array(variationSchema).defined(),
  })
  .defined()
  .camelCase()

export type ExperimentFull = yup.InferType<typeof experimentFullSchema>
