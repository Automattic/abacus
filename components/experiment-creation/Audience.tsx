import {
  FormControl,
  FormControlLabel,
  FormHelperText,
  FormLabel,
  InputAdornment,
  MenuItem,
  Radio,
  RadioGroup as MuiRadioGroup,
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableHead,
  TableRow,
  TextField as MuiTextField,
  Typography,
} from '@material-ui/core'
import { createStyles, makeStyles, Theme } from '@material-ui/core/styles'
import Autocomplete from '@material-ui/lab/Autocomplete'
import { Field, FormikProps, useField } from 'formik'
import { RadioGroup as FormikMuiRadioGroup, Select, TextField as FormikMuiTextField } from 'formik-material-ui'
import { AutocompleteProps, AutocompleteRenderInputParams, fieldToAutocomplete } from 'formik-material-ui-lab'
import _ from 'lodash'
import React, { useCallback, useState } from 'react'

import { PlatformToHuman } from '@/lib/experiments'
import { ExperimentFullNew, Platform, Segment, SegmentAssignmentNew, SegmentType, VariationNew } from '@/lib/schemas'
import { SegmentTypeToHuman } from '@/lib/segments'

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {},
    row: {
      margin: theme.spacing(6, 0),
      display: 'flex',
      alignItems: 'center',
    },
    segmentationHelperText: {},
    segmentationFieldSet: {
      width: '100%',
    },
    segmentationExclusionState: {
      display: 'flex',
      flexDirection: 'row',
      marginBottom: theme.spacing(1),
    },
    variationAllocatedPercentage: {
      width: '7rem',
    },
    variants: {
      width: 'auto',
    },
  }),
)

enum SegmentExclusionState {
  Exclude = 'exclude',
  Include = 'include',
}

const segments: Record<number, Segment> = {
  1: { segmentId: 1, name: 'us', type: SegmentType.Country },
  2: { segmentId: 2, name: 'au', type: SegmentType.Country },
  3: { segmentId: 3, name: 'en-US', type: SegmentType.Locale },
  4: { segmentId: 4, name: 'en-AU', type: SegmentType.Locale },
}

const SegmentsAutocomplete = (
  props: AutocompleteProps<Segment, true, false, false> & { segmentExclusionState: SegmentExclusionState },
) => {
  const {
    form: { setFieldValue },
    field: { name, value: outerValue },
    segmentExclusionState,
  } = props

  // Here we translate SegmentAssignment (outside) <-> Segment (inside)
  const segmentAssignmentToSegment = (segmentAssignment: SegmentAssignmentNew) => {
    const segment = segments[segmentAssignment.segmentId]
    /* istanbul ignore next; Should never happen */
    if (!segment) {
      throw new Error('Could not find segment with specified segmentId.')
    }
    return segment
  }
  const segmentToSegmentAssignment = useCallback(
    (segment: Segment): SegmentAssignmentNew => ({
      segmentId: segment.segmentId,
      isExcluded: segmentExclusionState === SegmentExclusionState.Exclude,
    }),
    [segmentExclusionState],
  )
  const onChange = useCallback(
    (_event, value: Segment[]) => {
      setFieldValue(name, value.map(segmentToSegmentAssignment))
    },
    [setFieldValue, name, segmentToSegmentAssignment],
  )
  const value = outerValue && (outerValue as SegmentAssignmentNew[]).map(segmentAssignmentToSegment)

  return (
    <Autocomplete
      {...fieldToAutocomplete(_.omit(props, 'segmentExclusionState'))}
      multiple={true}
      onChange={onChange}
      value={value}
      getOptionLabel={({ name, type }: Segment) => `${SegmentTypeToHuman[type]}: ${name}`}
    />
  )
}

const Audience = ({ formikProps }: { formikProps: FormikProps<{ experiment: Partial<ExperimentFullNew> }> }) => {
  const classes = useStyles()

  // The segmentExclusion code is currently split between here and SegmentAutocomplete
  // An improvement might be to have SegmentAutocomplete only handle Segment[] and for code here
  // to translate Segment <-> SegmentAssignment
  const [segmentAssignmentsField, _segmentAssignmentsFieldMeta, segmentAssignmentsFieldHelper] = useField(
    'experiment.segmentAssignments',
  )
  const [segmentExclusionState, setSegmentExclusionState] = useState<SegmentExclusionState>(
    SegmentExclusionState.Include,
  )
  const onChangeSegmentExclusionState = (event: React.SyntheticEvent<HTMLInputElement>, value: string) => {
    setSegmentExclusionState(value as SegmentExclusionState)
    segmentAssignmentsFieldHelper.setValue(
      (segmentAssignmentsField.value as SegmentAssignmentNew[]).map((segmentAssignment: SegmentAssignmentNew) => {
        return {
          ...segmentAssignment,
          isExcluded: value === SegmentExclusionState.Exclude,
        }
      }),
    )
  }

  return (
    <div className={classes.root}>
      <Typography variant='h2' gutterBottom>
        Audience
      </Typography>

      <div className={classes.row}>
        <FormControl component='fieldset'>
          <FormLabel required>Platform</FormLabel>
          <Field component={Select} name='experiment.platform'>
            {Object.values(Platform).map((platform) => (
              <MenuItem key={platform} value={platform}>
                {PlatformToHuman[platform]}
              </MenuItem>
            ))}
          </Field>
        </FormControl>
      </div>

      <div className={classes.row}>
        <FormControl component='fieldset'>
          <FormLabel required>User types</FormLabel>
          <FormHelperText>Types of users to include in experiment</FormHelperText>

          <Field component={FormikMuiRadioGroup} name='experiment.existingUsersAllowed' required>
            <FormControlLabel
              value='false'
              label='New users only'
              control={<Radio disabled={formikProps.isSubmitting} />}
              disabled={formikProps.isSubmitting}
            />
            <FormControlLabel
              value='true'
              label='All users (new + existing)'
              control={<Radio disabled={formikProps.isSubmitting} />}
              disabled={formikProps.isSubmitting}
            />
          </Field>
        </FormControl>
      </div>
      <div className={classes.row}>
        <FormControl component='fieldset' className={classes.segmentationFieldSet}>
          <FormLabel htmlFor='segments-select'>Targeting</FormLabel>
          <FormHelperText className={classes.segmentationHelperText}>
            Who should see this experiment? <br /> Add optional filters to include or exclude specific target audience
            segments.
          </FormHelperText>
          <MuiRadioGroup
            aria-label='include-or-exclude-segments'
            className={classes.segmentationExclusionState}
            value={segmentExclusionState}
            onChange={onChangeSegmentExclusionState}
          >
            <FormControlLabel
              value={SegmentExclusionState.Include}
              control={<Radio />}
              label='Include'
              name='non-formik-segment-exclusion-state-include'
            />
            <FormControlLabel
              value={SegmentExclusionState.Exclude}
              control={<Radio />}
              label='Exclude'
              name='non-formik-segment-exclusion-state-exclude'
            />
          </MuiRadioGroup>
          <Field
            name='experiment.segmentAssignments'
            component={SegmentsAutocomplete}
            options={Object.values(segments)}
            // TODO: Error state, see https://stackworx.github.io/formik-material-ui/docs/api/material-ui-lab
            renderInput={(params: AutocompleteRenderInputParams) => (
              <MuiTextField {...params} variant='outlined' placeholder='Search and select to customize' />
            )}
            segmentExclusionState={segmentExclusionState}
            fullWidth
            id='segments-select'
          />
        </FormControl>
      </div>
      <div className={classes.row}>
        <FormControl component='fieldset' className={classes.segmentationFieldSet}>
          <FormLabel htmlFor='variations-select'>Variations</FormLabel>
          <FormHelperText className={classes.segmentationHelperText}>
            Define the percentages to include in the experiment. <br /> Use &ldquo;control&rdquo; for the default
            (fallback) experience.
          </FormHelperText>
          <TableContainer>
            <Table className={classes.variants}>
              <TableHead>
                <TableRow>
                  <TableCell> Name </TableCell>
                  <TableCell> Allocated Percentage </TableCell>
                </TableRow>
              </TableHead>
              <TableBody>
                {(formikProps.values.experiment.variations as VariationNew[]).map((variation, idx) => {
                  return (
                    // The key here needs to be changed for variable variations
                    <TableRow key={variation.name}>
                      <TableCell>{variation.name}</TableCell>
                      <TableCell>
                        <Field
                          className={classes.variationAllocatedPercentage}
                          component={FormikMuiTextField}
                          name={`experiment.variations[${idx}].allocatedPercentage`}
                          type='number'
                          size='small'
                          variant='outlined'
                          inputProps={{ min: 1, max: 99 }}
                          required
                          InputProps={{
                            endAdornment: <InputAdornment position='end'>%</InputAdornment>,
                          }}
                        />
                      </TableCell>
                    </TableRow>
                  )
                })}
              </TableBody>
            </Table>
          </TableContainer>
        </FormControl>
      </div>
    </div>
  )
}

export default Audience
