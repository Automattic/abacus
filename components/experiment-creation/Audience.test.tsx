/* eslint-disable no-irregular-whitespace,@typescript-eslint/ban-ts-ignore */
import { fireEvent, render, screen, waitFor } from '@testing-library/react'
import { Formik, FormikProps } from 'formik'
import React from 'react'

import { createNewExperiment } from '@/lib/experiments'
import { ExperimentFullNew } from '@/lib/schemas'

import Audience from './Audience'

// Needed for testing the MuiCombobox
document.createRange = () => ({
  setStart: () => undefined,
  setEnd: () => undefined,
  // @ts-ignore; This is just for mocking
  commonAncestorContainer: {
    nodeName: 'BODY',
    ownerDocument: document,
  },
})

test('renders as expected', async () => {
  const { container } = render(
    <Formik
      initialValues={{ experiment: createNewExperiment() }}
      onSubmit={
        /* istanbul ignore next; This is unused */
        () => undefined
      }
    >
      {(formikProps: FormikProps<{ experiment: Partial<ExperimentFullNew> }>) => <Audience formikProps={formikProps} />}
    </Formik>,
  )
  expect(container).toMatchSnapshot()

  const segmentCombobox = screen.getByRole('combobox')
  fireEvent.click(segmentCombobox)

  const segmentComboboxInput = segmentCombobox.querySelector('input')

  // Typeguard
  if (!segmentComboboxInput) {
    throw new Error(`Can't find the segmentCombobox input`)
  }

  fireEvent.change(segmentComboboxInput, { target: { value: 'AU' } })

  await waitFor(() => screen.debug(screen.getByText(/AU/)))

  fireEvent.click(screen.getByText(/AU/))

  fireEvent.click(screen.getByLabelText(/Exclude/))
})
