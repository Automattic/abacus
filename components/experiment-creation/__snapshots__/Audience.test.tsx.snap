// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders as expected 1`] = `
<div>
  <div
    class="makeStyles-root-1"
  >
    <h2
      class="MuiTypography-root-7 MuiTypography-h2-13 MuiTypography-gutterBottom-27"
    >
      Audience
    </h2>
    <div
      class="makeStyles-row-2"
    >
      <fieldset
        class="MuiFormControl-root-37"
      >
        <label
          class="MuiFormLabel-root-41 MuiFormLabel-filled-46 MuiFormLabel-required-47"
        >
          Platform
          <span
            aria-hidden="true"
            class="MuiFormLabel-asterisk-48"
          >
             
            *
          </span>
        </label>
        <div
          class="MuiInputBase-root-73 MuiInput-root-59 MuiInput-underline-64 MuiInputBase-formControl-74 MuiInput-formControl-60"
        >
          <div
            aria-haspopup="listbox"
            aria-labelledby="mui-component-select-experiment.platform"
            class="MuiSelect-root-49 MuiSelect-select-50 MuiSelect-selectMenu-53 MuiInputBase-input-84 MuiInput-input-69"
            id="mui-component-select-experiment.platform"
            role="button"
            tabindex="0"
          >
            WordPress.com
          </div>
          <input
            name="experiment.platform"
            type="hidden"
            value="wpcom"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root-91 MuiSelect-icon-55"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
        </div>
      </fieldset>
    </div>
    <div
      class="makeStyles-row-2"
    >
      <fieldset
        class="MuiFormControl-root-37"
      >
        <label
          class="MuiFormLabel-root-41 MuiFormLabel-required-47"
        >
          User types
          <span
            aria-hidden="true"
            class="MuiFormLabel-asterisk-48"
          >
             
            *
          </span>
        </label>
        <p
          class="MuiFormHelperText-root-104"
        >
          Types of users to include in experiment
        </p>
        <div
          class="MuiFormGroup-root-112"
          inputlabelprops="[object Object]"
          required=""
          role="radiogroup"
        >
          <label
            class="MuiFormControlLabel-root-114"
          >
            <span
              aria-disabled="false"
              class="MuiButtonBase-root-138 MuiIconButton-root-129 PrivateSwitchBase-root-125 MuiRadio-root-120 MuiRadio-colorSecondary-124 MuiIconButton-colorSecondary-134"
            >
              <span
                class="MuiIconButton-label-137"
              >
                <input
                  class="PrivateSwitchBase-input-128"
                  name="experiment.existingUsersAllowed"
                  type="radio"
                  value="false"
                />
                <div
                  class="PrivateRadioButtonIcon-root-141"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root-91"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                    />
                  </svg>
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root-91 PrivateRadioButtonIcon-layer-142"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
                    />
                  </svg>
                </div>
              </span>
              <span
                class="MuiTouchRipple-root-217"
              />
            </span>
            <span
              class="MuiTypography-root-7 MuiFormControlLabel-label-119 MuiTypography-body1-9"
            >
              New users only
            </span>
          </label>
          <label
            class="MuiFormControlLabel-root-114"
          >
            <span
              aria-disabled="false"
              class="MuiButtonBase-root-138 MuiIconButton-root-129 PrivateSwitchBase-root-125 MuiRadio-root-120 MuiRadio-colorSecondary-124 MuiIconButton-colorSecondary-134"
            >
              <span
                class="MuiIconButton-label-137"
              >
                <input
                  class="PrivateSwitchBase-input-128"
                  name="experiment.existingUsersAllowed"
                  type="radio"
                  value="true"
                />
                <div
                  class="PrivateRadioButtonIcon-root-141"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root-91"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                    />
                  </svg>
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root-91 PrivateRadioButtonIcon-layer-142"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
                    />
                  </svg>
                </div>
              </span>
              <span
                class="MuiTouchRipple-root-217"
              />
            </span>
            <span
              class="MuiTypography-root-7 MuiFormControlLabel-label-119 MuiTypography-body1-9"
            >
              All users (new + existing)
            </span>
          </label>
        </div>
      </fieldset>
    </div>
    <div
      class="makeStyles-row-2"
    >
      <fieldset
        class="MuiFormControl-root-37 makeStyles-segmentationFieldSet-4"
      >
        <label
          class="MuiFormLabel-root-41"
          for="segments-select"
        >
          Targeting
        </label>
        <p
          class="MuiFormHelperText-root-104 makeStyles-segmentationHelperText-3"
        >
          Who should see this experiment?
          <br />
          Add optional filters to include or exclude specific target audience segments.
        </p>
        <div
          aria-label="include-or-exclude-segments"
          class="MuiFormGroup-root-112 makeStyles-segmentationExclusionState-5"
          role="radiogroup"
        >
          <label
            class="MuiFormControlLabel-root-114"
          >
            <span
              aria-disabled="false"
              class="MuiButtonBase-root-138 MuiIconButton-root-129 PrivateSwitchBase-root-125 MuiRadio-root-120 MuiRadio-colorSecondary-124 PrivateSwitchBase-checked-126 MuiRadio-checked-121 MuiIconButton-colorSecondary-134"
            >
              <span
                class="MuiIconButton-label-137"
              >
                <input
                  checked=""
                  class="PrivateSwitchBase-input-128"
                  name="mui-28273"
                  type="radio"
                  value="include"
                />
                <div
                  class="PrivateRadioButtonIcon-root-141 PrivateRadioButtonIcon-checked-143"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root-91"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                    />
                  </svg>
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root-91 PrivateRadioButtonIcon-layer-142"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
                    />
                  </svg>
                </div>
              </span>
              <span
                class="MuiTouchRipple-root-217"
              />
            </span>
            <span
              class="MuiTypography-root-7 MuiFormControlLabel-label-119 MuiTypography-body1-9"
            >
              Include
            </span>
          </label>
          <label
            class="MuiFormControlLabel-root-114"
          >
            <span
              aria-disabled="false"
              class="MuiButtonBase-root-138 MuiIconButton-root-129 PrivateSwitchBase-root-125 MuiRadio-root-120 MuiRadio-colorSecondary-124 MuiIconButton-colorSecondary-134"
            >
              <span
                class="MuiIconButton-label-137"
              >
                <input
                  class="PrivateSwitchBase-input-128"
                  name="mui-28273"
                  type="radio"
                  value="exclude"
                />
                <div
                  class="PrivateRadioButtonIcon-root-141"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root-91"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                    />
                  </svg>
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root-91 PrivateRadioButtonIcon-layer-142"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
                    />
                  </svg>
                </div>
              </span>
              <span
                class="MuiTouchRipple-root-217"
              />
            </span>
            <span
              class="MuiTypography-root-7 MuiFormControlLabel-label-119 MuiTypography-body1-9"
            >
              Exclude
            </span>
          </label>
        </div>
        <div
          aria-expanded="false"
          class="MuiAutocomplete-root-144 MuiAutocomplete-fullWidth-145 MuiAutocomplete-hasClearIcon-150 MuiAutocomplete-hasPopupIcon-149"
          name="experiment.segmentAssignments"
          role="combobox"
          segmentexclusionstate="include"
        >
          <div
            class="MuiFormControl-root-37 MuiTextField-root-168 MuiFormControl-fullWidth-40"
          >
            <div
              class="MuiInputBase-root-73 MuiOutlinedInput-root-169 MuiAutocomplete-inputRoot-151 MuiInputBase-fullWidth-83 MuiInputBase-formControl-74 MuiInputBase-adornedEnd-78 MuiOutlinedInput-adornedEnd-174"
            >
              <input
                aria-autocomplete="list"
                aria-invalid="false"
                autocapitalize="none"
                autocomplete="off"
                class="MuiInputBase-input-84 MuiOutlinedInput-input-179 MuiAutocomplete-input-152 MuiAutocomplete-inputFocused-153 MuiInputBase-inputAdornedEnd-89 MuiOutlinedInput-inputAdornedEnd-183"
                id="segments-select"
                placeholder="Search and select to customize"
                spellcheck="false"
                type="text"
                value=""
              />
              <div
                class="MuiAutocomplete-endAdornment-154"
              >
                <button
                  aria-label="Clear"
                  class="MuiButtonBase-root-138 MuiIconButton-root-129 MuiAutocomplete-clearIndicator-155"
                  tabindex="-1"
                  title="Clear"
                  type="button"
                >
                  <span
                    class="MuiIconButton-label-137"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root-91 MuiSvgIcon-fontSizeSmall-98"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                      />
                    </svg>
                  </span>
                  <span
                    class="MuiTouchRipple-root-217"
                  />
                </button>
                <button
                  aria-label="Open"
                  class="MuiButtonBase-root-138 MuiIconButton-root-129 MuiAutocomplete-popupIndicator-157"
                  tabindex="-1"
                  title="Open"
                  type="button"
                >
                  <span
                    class="MuiIconButton-label-137"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root-91"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M7 10l5 5 5-5z"
                      />
                    </svg>
                  </span>
                  <span
                    class="MuiTouchRipple-root-217"
                  />
                </button>
              </div>
              <fieldset
                aria-hidden="true"
                class="PrivateNotchedOutline-root-184 MuiOutlinedInput-notchedOutline-178"
                style="padding-left: 8px;"
              >
                <legend
                  class="PrivateNotchedOutline-legend-185"
                  style="width: 0.01px;"
                >
                  <span>
                    ​
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
      </fieldset>
    </div>
    <div
      class="makeStyles-row-2"
    >
      <fieldset
        class="MuiFormControl-root-37 makeStyles-segmentationFieldSet-4"
      >
        <label
          class="MuiFormLabel-root-41"
          for="variations-select"
        >
          Variations
        </label>
        <p
          class="MuiFormHelperText-root-104 makeStyles-segmentationHelperText-3"
        >
          Define the percentages to include in the experiment.
          <br />
          Use “control” for the default (fallback) experience.
        </p>
        <div
          class="MuiTableContainer-root-188"
        >
          <table
            class="MuiTable-root-189"
          >
            <thead
              class="MuiTableHead-root-191"
            >
              <tr
                class="MuiTableRow-root-192 MuiTableRow-head-195"
              >
                <th
                  class="MuiTableCell-root-197 MuiTableCell-head-198"
                  scope="col"
                >
                   Name 
                </th>
                <th
                  class="MuiTableCell-root-197 MuiTableCell-head-198"
                  scope="col"
                >
                   Allocated Percentage 
                </th>
              </tr>
            </thead>
            <tbody
              class="MuiTableBody-root-209"
            >
              <tr
                class="MuiTableRow-root-192"
              >
                <td
                  class="MuiTableCell-root-197 MuiTableCell-body-199"
                >
                  control
                </td>
                <td
                  class="MuiTableCell-root-197 MuiTableCell-body-199"
                >
                  <div
                    class="MuiFormControl-root-37 MuiTextField-root-168 makeStyles-variationAllocatedPercentage-6"
                  >
                    <div
                      class="MuiInputBase-root-73 MuiOutlinedInput-root-169 MuiInputBase-formControl-74 MuiInputBase-adornedEnd-78 MuiOutlinedInput-adornedEnd-174 MuiInputBase-marginDense-80 MuiOutlinedInput-marginDense-176"
                    >
                      <input
                        aria-invalid="false"
                        class="MuiInputBase-input-84 MuiOutlinedInput-input-179 MuiInputBase-inputAdornedEnd-89 MuiOutlinedInput-inputAdornedEnd-183 MuiInputBase-inputMarginDense-85 MuiOutlinedInput-inputMarginDense-180"
                        max="99"
                        min="1"
                        name="experiment.variations[0].allocatedPercentage"
                        required=""
                        type="number"
                        value="50"
                      />
                      <div
                        class="MuiInputAdornment-root-210 MuiInputAdornment-positionEnd-213 MuiInputAdornment-marginDense-216"
                      >
                        <p
                          class="MuiTypography-root-7 MuiTypography-body1-9 MuiTypography-colorTextSecondary-33"
                        >
                          %
                        </p>
                      </div>
                      <fieldset
                        aria-hidden="true"
                        class="PrivateNotchedOutline-root-184 MuiOutlinedInput-notchedOutline-178"
                        style="padding-left: 8px;"
                      >
                        <legend
                          class="PrivateNotchedOutline-legend-185"
                          style="width: 0.01px;"
                        >
                          <span>
                            ​
                          </span>
                        </legend>
                      </fieldset>
                    </div>
                  </div>
                </td>
              </tr>
              <tr
                class="MuiTableRow-root-192"
              >
                <td
                  class="MuiTableCell-root-197 MuiTableCell-body-199"
                >
                  treatment
                </td>
                <td
                  class="MuiTableCell-root-197 MuiTableCell-body-199"
                >
                  <div
                    class="MuiFormControl-root-37 MuiTextField-root-168 makeStyles-variationAllocatedPercentage-6"
                  >
                    <div
                      class="MuiInputBase-root-73 MuiOutlinedInput-root-169 MuiInputBase-formControl-74 MuiInputBase-adornedEnd-78 MuiOutlinedInput-adornedEnd-174 MuiInputBase-marginDense-80 MuiOutlinedInput-marginDense-176"
                    >
                      <input
                        aria-invalid="false"
                        class="MuiInputBase-input-84 MuiOutlinedInput-input-179 MuiInputBase-inputAdornedEnd-89 MuiOutlinedInput-inputAdornedEnd-183 MuiInputBase-inputMarginDense-85 MuiOutlinedInput-inputMarginDense-180"
                        max="99"
                        min="1"
                        name="experiment.variations[1].allocatedPercentage"
                        required=""
                        type="number"
                        value="50"
                      />
                      <div
                        class="MuiInputAdornment-root-210 MuiInputAdornment-positionEnd-213 MuiInputAdornment-marginDense-216"
                      >
                        <p
                          class="MuiTypography-root-7 MuiTypography-body1-9 MuiTypography-colorTextSecondary-33"
                        >
                          %
                        </p>
                      </div>
                      <fieldset
                        aria-hidden="true"
                        class="PrivateNotchedOutline-root-184 MuiOutlinedInput-notchedOutline-178"
                        style="padding-left: 8px;"
                      >
                        <legend
                          class="PrivateNotchedOutline-legend-185"
                          style="width: 0.01px;"
                        >
                          <span>
                            ​
                          </span>
                        </legend>
                      </fieldset>
                    </div>
                  </div>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </fieldset>
    </div>
  </div>
</div>
`;
